lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -1652,6 +1670,77 @@ reapplyFunctionConversion(SILValue newFunc, SILValue oldFunc,
   llvm_unreachable("Unhandled function convertion instruction");
 }
 
+/// Emits a reference to a VJP corresponding to `original`, differentiated with
+/// `indices`. On failure, returns a null `SILValue`.
+///
+/// Creates new differentiation tasks, if necessary, using `invoker` as the
+/// invoker. Calls `taskCallback` for all newly-created tasks (but may also call
+/// `taskCallback` for already-existing tasks), so that the caller can make sure
+/// that the task actually gets executed.
+static SILValue emitVJPReference(
+    ADContext &context, SILBuilder &builder, SILAutoDiffIndices indices,
+    SILValue original, DifferentiationInvoker invoker,
+    std::function<void(DifferentiationTask *)> taskCallback) {
+  if (auto *originalFRI = findReferenceToVisibleFunction(original)) {
+    auto loc = originalFRI->getLoc();
+    auto *originalFn = originalFRI->getReferencedFunction();
+    auto *task = context.lookUpOrRegisterDifferentiationTask(originalFn,
+                                                             indices, invoker);
+    taskCallback(task);
+    auto *vjpRef = builder.createFunctionRef(loc, task->getVJP());
+    return reapplyFunctionConversion(vjpRef, originalFRI, original, builder,
+                                     loc);
+  }
+  if (auto *witnessMethod = dyn_cast<WitnessMethodInst>(original)) {
```

Comment:
I think this will get a lot more cases quite soon. If I do the recursive thing that I mentioned below, it'll handle a bunch of function conversion instructions. Also, this will eventually need to handle `apply` instructions, where the `apply` creates a function whose VJP we want.